### ISSUE TITLE
Hash transform and the ranking sample

### ISSUE THREAD

**Issue by TomFinley** [microsoft/NimbusML#59]
We have this sample here.

https://github.com/Microsoft/NimbusML-Samples/blob/master/samples/2.5%20%5BNumeric%5D%20Learning-to-Rank%20with%20Microsoft%20Bing%20Data.ipynb

The sample is incorrect because it suggests using `ToKey` to transform group IDs into keys, but this builds a dictionary. Since group IDs in the *test set* will almost certainly not resemble group IDs in the *train set*, the mapping on the test set will produce the "missing key value" as the group key, leading to the evaluator saying "hey these all have the same key," so the entire dataset gets collectively evaluated as if it were a single query, which in turn leads to nonsensical NDCG numbers (though, the scores of the documents themselves will not be impacted).

The advice for ML.NET side is to not use that -- instead we suggest the other common way of turning things in to keys, that is, the hash, which does *not* depend on seeing that key in the training set, allowing it to still work. Indeed, I see the entry-point for hash was published here:

https://github.com/Microsoft/NimbusML/blob/e1004720ec0c252ba87f02c190c33739d9c00f20/src/python/nimbusml/internal/entrypoints/transforms_hashconverter.py

However, it looks like the step where actually turned this into something accessible by users was overlooked.

What raises this to the dignity of a bug I think is that we've explicitly identified ranking as a targeted scenario, yet we've forgotten the key thing that will make evaluation actually work, possibly by not realizing that `ToKey` produces useless evaluation in this case.

**Comment by montebhoover** [comment]
Thanks Tom.  So I can add a note of clarification on when to use one versus the other, what are the common cases where `ToKey` is more appropriate than `HashConverter`?

**Comment by justinormont** [comment]
@TomFinley: Wow, very good catch.

**Comment by justinormont** [comment]
@montebhoover: `ToKey` is mainly used when converting the Label column from `Text` to `Key` formats. The use case is akin to scikit-learn's [LabelEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html).
